crates/driver/src/main.rs

```diff
@@ -59,24 +60,111 @@ pub async fn run(
+                        (None, Some(private_key)) => ethcontract::Account::Offline(
+                            ethcontract::PrivateKey::from_hex_str(private_key)
+                                .expect("a valid private key in --solver-private-key"),
+                            None,
+                        ),
+                        _ => panic!(
+                            "exactly one of --solver-address, --solver-private-key must be \
+                             specified",
+                        ),
+                    },
+                    eth: eth.clone(),
+                    pool: tx_pool.clone(),
+                    gas_price_calculation: args
+                        .submission
+                        .submission_gas_price_calculation
+                        .iter()
+                        .map(|calculation| match calculation {
+                            cli::GasPriceCalculation::EthGasStation => {
+                                mempool::GasPriceCalculation::EthGasStation
+                            }
+                            cli::GasPriceCalculation::GasNow => {
+                                mempool::GasPriceCalculation::GasNow
+                            }
+                            cli::GasPriceCalculation::GnosisSafe => {
+                                mempool::GasPriceCalculation::GnosisSafe
+                            }
+                            cli::GasPriceCalculation::Web3 => mempool::GasPriceCalculation::Web3,
+                            cli::GasPriceCalculation::BlockNative => {
+                                mempool::GasPriceCalculation::BlockNative {
+                                    api_key: args.blocknative_api_key.clone().expect(
+                                        "BlockNative price calculation requires \
+                                         --blocknative-api-key",
+                                    ),
+                                }
+                            }
+                            cli::GasPriceCalculation::Native => {
+                                mempool::GasPriceCalculation::Native
+                            }
+                        })
+                        .collect(),
+                };
+                match mempool {
+                    cli::Mempool::Public => vec![Mempool::public(
+                        config,
+                        if args
+                            .submission
+                            .submission_disable_high_risk_public_mempool_transactions
+                        {
+                            mempool::HighRisk::Disabled
+                        } else {
+                            mempool::HighRisk::Enabled
+                        },
+                    )
+                    .await
+                    .unwrap()],
+                    cli::Mempool::Flashbots => {
+                        join_all(args.flashbots_api_urls.iter().map(|url| async {
+                            Mempool::flashbots(
```

Comment:
When creating a flashbots instance a new `dyn GasPriceEstimating` also gets created although one could only have a single instance and copy that around where it's needed with the `Arc`. Having only one instance of `dyn GasPriceEstimating` used everywhere could reduce latency by sharing or caching requests.
Just wanted to bring it up in case that was not intended.

Reply:
I do think the `dyn GasPriceEstimating` should be shared if possible. Basically it polls `eth_getFeeHistory` and updates its internal state based on the resulting data, so having multiple instances of it are a bit wasteful.